File: leetcode/day10/cpp/so.cpp
#include "so.h"
#include <iostream>

int Solution::maxSubArray(std::vector<int> &nums)
{

    std::vector<int> dp = {nums[0]};

    for (size_t i = 1; i < nums.size(); i++)
    {
        /* code */
        auto num = nums[i];

        if (dp[i - 1] > 0)
        {

            dp.push_back(dp[i - 1] + nums[i]);
        }
        else
        {
            dp.push_back(num);
        }
    }

    int res = dp[0];

    for (size_t i = 0; i < dp.size(); i++)
    {
        /* code */
        res = std::max(res, dp[i]);
    }

    return res;
}

File: leetcode/day10/python/so.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        dp[0] = nums[0]

        for i, num in enumerate(nums):
            if (i == 0):
                continue
            if dp[i-1] > 0:
                dp[i] = dp[i-1] + num
            else:
                dp[i] = num

        res = dp[0]

        for r in dp:
            res = max(r, res)
        return res
